File: packages/csharp/src/V2/TransactionsClient.cs
using System;
using System.Collections.Generic;
using System.Threading;
using System.Threading.Tasks;
using Tesote.Sdk.Internal;
using Tesote.Sdk.Models;

namespace Tesote.Sdk.V2;

/// <summary>
/// v2 Transactions resource. Per-account list/export/sync, plus the cross-account
/// lookup, bulk fetch, and search endpoints.
/// </summary>
public sealed class TransactionsClient
{
    private readonly Transport _transport;

    internal TransactionsClient(Transport transport)
    {
        _transport = transport;
    }

    /// <summary>Filter parameters accepted by the v2 transactions list/search endpoints.</summary>
    public sealed class ListFilters
    {
        /// <summary>ISO8601 date — earliest tesote-imported date to include.</summary>
        public string? StartDate { get; set; }
        /// <summary>ISO8601 date — latest tesote-imported date to include.</summary>
        public string? EndDate { get; set; }
        /// <summary>Server-side scope name (per docs).</summary>
        public string? Scope { get; set; }
        /// <summary>1-indexed page number.</summary>
        public int? Page { get; set; }
        /// <summary>Page size, default 50, max 100.</summary>
        public int? PerPage { get; set; }
        /// <summary>Cursor — fetch transactions after this id.</summary>
        public string? TransactionsAfterId { get; set; }
        /// <summary>Cursor — fetch transactions before this id.</summary>
        public string? TransactionsBeforeId { get; set; }
        /// <summary>ISO8601 — transaction-date floor.</summary>
        public string? TransactionDateAfter { get; set; }
        /// <summary>ISO8601 — transaction-date ceiling.</summary>
        public string? TransactionDateBefore { get; set; }
        /// <summary>ISO8601 — server-side <c>created_at</c> floor.</summary>
        public string? CreatedAfter { get; set; }
        /// <summary>ISO8601 — server-side <c>updated_at</c> floor.</summary>
        public string? UpdatedAfter { get; set; }
        /// <summary>Numeric — minimum amount (signed).</summary>
        public decimal? AmountMin { get; set; }
        /// <summary>Numeric — maximum amount.</summary>
        public decimal? AmountMax { get; set; }
        /// <summary>Numeric — exact amount match.</summary>
        public decimal? Amount { get; set; }
        /// <summary>Status filter (e.g. <c>posted</c>, <c>pending</c>).</summary>
        public string? Status { get; set; }
        /// <summary>Filter by category id.</summary>
        public string? CategoryId { get; set; }
        /// <summary>Filter by counterparty id.</summary>
        public string? CounterpartyId { get; set; }
        /// <summary>Search string (description / counterparty name).</summary>
        public string? Q { get; set; }
        /// <summary>Type filter (per docs).</summary>
        public string? Type { get; set; }
        /// <summary>Reference-code filter.</summary>
        public string? ReferenceCode { get; set; }

        internal IReadOnlyDictionary<string, string>? Build()
        {
            return new QueryBuilder()
                .Add("start_date", StartDate)
                .Add("end_date", EndDate)
                .Add("scope", Scope)
                .Add("page", Page)
                .Add("per_page", PerPage)
                .Add("transactions_after_id", TransactionsAfterId)
                .Add("transactions_before_id", TransactionsBeforeId)
                .Add("transaction_date_after", TransactionDateAfter)
                .Add("transaction_date_before", TransactionDateBefore)
                .Add("created_after", CreatedAfter)
                .Add("updated_after", UpdatedAfter)
                .Add("amount_min", AmountMin)
                .Add("amount_max", AmountMax)
                .Add("amount", Amount)
                .Add("status", Status)
                .Add("category_id", CategoryId)
                .Add("counterparty_id", CounterpartyId)
                .Add("q", Q)
                .Add("type", Type)
                .Add("reference_code", ReferenceCode)
                .BuildOrNull();
        }
    }

    /// <summary>List transactions for an account with the full v2 filter surface.</summary>
    public async Task<TransactionListResponse> ListForAccountAsync(
        string accountId,
        ListFilters? filters = null,
        CancellationToken ct = default)
    {
        ArgumentException.ThrowIfNullOrEmpty(accountId);
        var path = "/v2/accounts/" + Uri.EscapeDataString(accountId) + "/transactions";
        var opts = RequestOptions.Get(path);
        opts.Query = (filters ?? new ListFilters()).Build();
        opts.CacheTtl = TimeSpan.FromMinutes(1);
        var node = await _transport.RequestAsync(opts, ct).ConfigureAwait(false);
        return Json.Deserialize<TransactionListResponse>(node);
    }

    /// <summary>Export transactions as CSV or JSON file. Returns raw bytes plus content-type.</summary>
    public Task<RawResponse> ExportAsync(
        string accountId,
        string format = "csv",
        ListFilters? filters = null,
        CancellationToken ct = default)
    {
        ArgumentException.ThrowIfNullOrEmpty(accountId);
        ArgumentException.ThrowIfNullOrEmpty(format);
        var path = "/v2/accounts/" + Uri.EscapeDataString(accountId) + "/transactions/export";
        var query = (filters ?? new ListFilters()).Build();
        var qb = new QueryBuilder().Add("format", format);
        if (query is not null)
        {
            foreach (var kv in query)
            {
                qb.Add(kv.Key, kv.Value);
            }
        }
        var opts = RequestOptions.Get(path);
        opts.Query = qb.BuildOrNull();
        return _transport.RequestRawAsync(opts, ct);
    }

    /// <summary>Sync request body for POST /v2/.../transactions/sync.</summary>
    public sealed class SyncRequest
    {
        /// <summary>Number of transactions to fetch (1..1000).</summary>
        public int? Count { get; set; }
        /// <summary>Opaque cursor or the literal <c>"now"</c>.</summary>
        public string? Cursor { get; set; }
        /// <summary>Optional flags accepted by the API.</summary>
        public SyncOptions? Options { get; set; }
    }

    /// <summary>Optional flag bag for <see cref="SyncRequest"/>.</summary>
    public sealed class SyncOptions
    {
        /// <summary>Include per-transaction running balances when the workspace allows it.</summary>
        public bool? IncludeRunningBalance { get; set; }
    }

    /// <summary>Cursor-based incremental sync for a single account.</summary>
    public async Task<SyncResult> SyncAsync(
        string accountId,
        SyncRequest request,
        string? idempotencyKey = null,
        CancellationToken ct = default)
    {
        ArgumentException.ThrowIfNullOrEmpty(accountId);
        ArgumentNullException.ThrowIfNull(request);
        var path = "/v2/accounts/" + Uri.EscapeDataString(accountId) + "/transactions/sync";
        var body = BuildSyncBody(request);
        var opts = Requests.Json("POST", path, body, idempotencyKey);
        var node = await _transport.RequestAsync(opts, ct).ConfigureAwait(false);
        return Json.Deserialize<SyncResult>(node);
    }

    /// <summary>Legacy non-nested sync route kept for backwards compatibility.</summary>
    public async Task<SyncResult> SyncLegacyAsync(
        SyncRequest request,
        string? idempotencyKey = null,
        CancellationToken ct = default)
    {
        ArgumentNullException.ThrowIfNull(request);
        var body = BuildSyncBody(request);
        var opts = Requests.Json("POST", "/v2/transactions/sync", body, idempotencyKey);
        var node = await _transport.RequestAsync(opts, ct).ConfigureAwait(false);
        return Json.Deserialize<SyncResult>(node);
    }

    /// <summary>Fetch a single transaction by id (v1-shape payload).</summary>
    public async Task<Transaction> GetAsync(string transactionId, CancellationToken ct = default)
    {
        ArgumentException.ThrowIfNullOrEmpty(transactionId);
        var opts = RequestOptions.Get("/v2/transactions/" + Uri.EscapeDataString(transactionId));
        opts.CacheTtl = TimeSpan.FromMinutes(5);
        var node = await _transport.RequestAsync(opts, ct).ConfigureAwait(false);
        return Json.Deserialize<Transaction>(node);
    }

    /// <summary>Bulk-fetch transactions for up to 100 accounts in one call.</summary>
    public async Task<BulkResponse> BulkAsync(
        IReadOnlyList<string> accountIds,
        int? page = null,
        int? perPage = null,
        int? limit = null,
        int? offset = null,
        string? idempotencyKey = null,
        CancellationToken ct = default)
    {
        ArgumentNullException.ThrowIfNull(accountIds);
        var body = new Dictionary<string, object?>
        {
            ["account_ids"] = accountIds,
            ["page"] = page,
            ["per_page"] = perPage,
            ["limit"] = limit,
            ["offset"] = offset,
        };
        var opts = Requests.Json("POST", "/v2/transactions/bulk", body, idempotencyKey);
        var node = await _transport.RequestAsync(opts, ct).ConfigureAwait(false);
        return Json.Deserialize<BulkResponse>(node);
    }

    /// <summary>Search transactions across accounts using a substring match plus the standard filter set.</summary>
    public async Task<SearchResult> SearchAsync(
        string query,
        string? accountId = null,
        int? limit = null,
        int? offset = null,
        ListFilters? filters = null,
        CancellationToken ct = default)
    {
        ArgumentException.ThrowIfNullOrEmpty(query);
        var qb = new QueryBuilder()
            .Add("q", query)
            .Add("account_id", accountId)
            .Add("limit", limit)
            .Add("offset", offset);
        var extra = (filters ?? new ListFilters()).Build();
        if (extra is not null)
        {
            foreach (var kv in extra)
            {
                if (!kv.Key.Equals("q", StringComparison.Ordinal))
                {
                    qb.Add(kv.Key, kv.Value);
                }
            }
        }
        var opts = RequestOptions.Get("/v2/transactions/search");
        opts.Query = qb.BuildOrNull();
        var node = await _transport.RequestAsync(opts, ct).ConfigureAwait(false);
        return Json.Deserialize<SearchResult>(node);
    }

    private static Dictionary<string, object?> BuildSyncBody(SyncRequest request)
    {
        var body = new Dictionary<string, object?>();
        if (request.Count is not null)
        {
            body["count"] = request.Count.Value;
        }
        if (request.Cursor is not null)
        {
            body["cursor"] = request.Cursor;
        }
        if (request.Options is not null)
        {
            var inner = new Dictionary<string, object?>();
            if (request.Options.IncludeRunningBalance is not null)
            {
                inner["include_running_balance"] = request.Options.IncludeRunningBalance.Value;
            }
            body["options"] = inner;
        }
        return body;
    }
}

File: packages/java/src/main/java/com/tesote/sdk/v2/TransactionsClient.java
package com.tesote.sdk.v2;

import com.fasterxml.jackson.databind.JsonNode;
import com.tesote.sdk.Transport;
import com.tesote.sdk.internal.Json;
import com.tesote.sdk.internal.QueryParams;
import com.tesote.sdk.models.BulkResponse;
import com.tesote.sdk.models.SyncTransactionsResponse;
import com.tesote.sdk.models.Transaction;
import com.tesote.sdk.models.TransactionsExport;
import com.tesote.sdk.models.TransactionsPage;
import com.tesote.sdk.models.TransactionsSearchResponse;

import java.time.Duration;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;

/**
 * v2 transactions: nested list, sync (cursor + legacy), get, bulk read,
 * search, export.
 */
public final class TransactionsClient {
    private static final Duration GET_CACHE = Duration.ofMinutes(5);
    private static final Duration LIST_CACHE = Duration.ofMinutes(1);

    private final Transport transport;

    public TransactionsClient(Transport transport) { this.transport = transport; }

    public TransactionsPage listForAccount(String accountId) {
        return listForAccount(accountId, new ListParams());
    }

    public TransactionsPage listForAccount(String accountId, ListParams params) {
        Objects.requireNonNull(accountId, "accountId");
        Objects.requireNonNull(params, "params");
        Transport.Options opts = Transport.Options.get(
                "/v2/accounts/" + AccountsClient.encode(accountId) + "/transactions")
                .query(params.toQuery())
                .cacheTtl(LIST_CACHE);
        return Json.treeToValue(transport.request(opts), TransactionsPage.class);
    }

    public Transaction get(String id) {
        Objects.requireNonNull(id, "id");
        Transport.Options opts = Transport.Options.get(
                "/v2/transactions/" + AccountsClient.encode(id))
                .cacheTtl(GET_CACHE);
        return Json.treeToValue(transport.request(opts), Transaction.class);
    }

    /** Cursor-based incremental sync scoped to one account. */
    public SyncTransactionsResponse sync(String accountId, SyncRequest request) {
        Objects.requireNonNull(accountId, "accountId");
        Objects.requireNonNull(request, "request");
        Transport.Options opts = new Transport.Options();
        opts.method = "POST";
        opts.path = "/v2/accounts/" + AccountsClient.encode(accountId) + "/transactions/sync";
        opts.jsonBody(request.toMap());
        return Json.treeToValue(transport.request(opts), SyncTransactionsResponse.class);
    }

    /** Legacy non-nested sync. Caller specifies account context inside the body. */
    public SyncTransactionsResponse syncLegacy(SyncRequest request) {
        Objects.requireNonNull(request, "request");
        Transport.Options opts = new Transport.Options();
        opts.method = "POST";
        opts.path = "/v2/transactions/sync";
        opts.jsonBody(request.toMap());
        return Json.treeToValue(transport.request(opts), SyncTransactionsResponse.class);
    }

    /** Fetch transactions across multiple accounts in one round-trip. */
    public BulkResponse bulk(BulkRequest request) {
        Objects.requireNonNull(request, "request");
        Transport.Options opts = new Transport.Options();
        opts.method = "POST";
        opts.path = "/v2/transactions/bulk";
        opts.jsonBody(request.toMap());
        return Json.treeToValue(transport.request(opts), BulkResponse.class);
    }

    public TransactionsSearchResponse search(SearchParams params) {
        Objects.requireNonNull(params, "params");
        Transport.Options opts = Transport.Options.get("/v2/transactions/search")
                .query(params.toQuery());
        return Json.treeToValue(transport.request(opts), TransactionsSearchResponse.class);
    }

    /**
     * Download a CSV or JSON export of transactions for an account. Returns
     * the raw bytes plus the server-provided filename when present.
     */
    public TransactionsExport export(String accountId, ExportParams params) {
        Objects.requireNonNull(accountId, "accountId");
        Objects.requireNonNull(params, "params");
        Transport.Options opts = new Transport.Options();
        opts.method = "GET";
        opts.path = "/v2/accounts/" + AccountsClient.encode(accountId) + "/transactions/export";
        opts.query = params.toQuery();
        Transport.RawResponse raw = transport.requestRaw(opts);
        return new TransactionsExport(
                raw.body(),
                parseFilename(raw.contentDisposition()),
                raw.contentType());
    }

    private static String parseFilename(String contentDisposition) {
        if (contentDisposition == null) return null;
        // why: "attachment; filename=foo.csv" or filename*= variants.
        for (String part : contentDisposition.split(";")) {
            String trim = part.trim();
            if (trim.startsWith("filename=")) {
                String v = trim.substring("filename=".length()).trim();
                if (v.startsWith("\"") && v.endsWith("\"") && v.length() >= 2) {
                    v = v.substring(1, v.length() - 1);
                }
                return v;
            }
        }
        return null;
    }

    /** Body for sync / syncLegacy. */
    public static final class SyncRequest {
        public Integer count;
        public String cursor;
        public Boolean includeRunningBalance;

        public SyncRequest count(int v) { this.count = v; return this; }
        public SyncRequest cursor(String v) { this.cursor = v; return this; }
        public SyncRequest includeRunningBalance(boolean v) { this.includeRunningBalance = v; return this; }

        Map<String, Object> toMap() {
            Map<String, Object> m = new LinkedHashMap<>();
            if (count != null) m.put("count", count);
            if (cursor != null) m.put("cursor", cursor);
            if (includeRunningBalance != null) {
                Map<String, Object> opts = new LinkedHashMap<>();
                opts.put("include_running_balance", includeRunningBalance);
                m.put("options", opts);
            }
            return m;
        }
    }

    /** Body for {@link #bulk(BulkRequest)}. */
    public static final class BulkRequest {
        public List<String> accountIds;
        public Integer page;
        public Integer perPage;
        public Integer limit;
        public Integer offset;

        public BulkRequest accountIds(List<String> v) { this.accountIds = v; return this; }
        public BulkRequest page(int v) { this.page = v; return this; }
        public BulkRequest perPage(int v) { this.perPage = v; return this; }
        public BulkRequest limit(int v) { this.limit = v; return this; }
        public BulkRequest offset(int v) { this.offset = v; return this; }

        Map<String, Object> toMap() {
            Map<String, Object> m = new LinkedHashMap<>();
            if (accountIds != null) m.put("account_ids", accountIds);
            if (page != null) m.put("page", page);
            if (perPage != null) m.put("per_page", perPage);
            if (limit != null) m.put("limit", limit);
            if (offset != null) m.put("offset", offset);
            return m;
        }
    }

    /** Search filters; reuses the listForAccount filter shape. */
    public static final class SearchParams {
        public String q;
        public String accountId;
        public Integer limit;
        public Integer offset;
        public ListParams filters = new ListParams();

        public SearchParams q(String v) { this.q = v; return this; }
        public SearchParams accountId(String v) { this.accountId = v; return this; }
        public SearchParams limit(int v) { this.limit = v; return this; }
        public SearchParams offset(int v) { this.offset = v; return this; }
        public SearchParams filters(ListParams v) { this.filters = v; return this; }

        Map<String, String> toQuery() {
            Map<String, String> base = filters == null ? Map.of() : filters.toQuery();
            Map<String, String> result = new LinkedHashMap<>(base);
            QueryParams qp = QueryParams.of()
                    .put("q", q)
                    .put("account_id", accountId)
                    .put("limit", limit)
                    .put("offset", offset);
            qp.build().forEach(result::put);
            return result;
        }
    }

    /** Export query params: list filters plus {@code format}. */
    public static final class ExportParams {
        public ListParams filters = new ListParams();
        public TransactionsExport.Format format;

        public ExportParams filters(ListParams v) { this.filters = v; return this; }
        public ExportParams format(TransactionsExport.Format f) { this.format = f; return this; }

        Map<String, String> toQuery() {
            Map<String, String> base = filters == null ? Map.of() : filters.toQuery();
            Map<String, String> result = new LinkedHashMap<>(base);
            if (format != null) result.put("format", format.wire());
            return result;
        }
    }

    /** Full v2 transaction list filter set (used by list, search, export). */
    public static final class ListParams {
        public String startDate;
        public String endDate;
        public String scope;
        public Integer page;
        public Integer perPage;
        public String transactionsAfterId;
        public String transactionsBeforeId;
        public String transactionDateAfter;
        public String transactionDateBefore;
        public String createdAfter;
        public String updatedAfter;
        public String amountMin;
        public String amountMax;
        public String amount;
        public String status;
        public String categoryId;
        public String counterpartyId;
        public String q;
        public String type;
        public String referenceCode;

        public ListParams startDate(String v) { this.startDate = v; return this; }
        public ListParams endDate(String v) { this.endDate = v; return this; }
        public ListParams scope(String v) { this.scope = v; return this; }
        public ListParams page(int v) { this.page = v; return this; }
        public ListParams perPage(int v) { this.perPage = v; return this; }
        public ListParams transactionsAfterId(String v) { this.transactionsAfterId = v; return this; }
        public ListParams transactionsBeforeId(String v) { this.transactionsBeforeId = v; return this; }
        public ListParams transactionDateAfter(String v) { this.transactionDateAfter = v; return this; }
        public ListParams transactionDateBefore(String v) { this.transactionDateBefore = v; return this; }
        public ListParams createdAfter(String v) { this.createdAfter = v; return this; }
        public ListParams updatedAfter(String v) { this.updatedAfter = v; return this; }
        public ListParams amountMin(String v) { this.amountMin = v; return this; }
        public ListParams amountMax(String v) { this.amountMax = v; return this; }
        public ListParams amount(String v) { this.amount = v; return this; }
        public ListParams status(String v) { this.status = v; return this; }
        public ListParams categoryId(String v) { this.categoryId = v; return this; }
        public ListParams counterpartyId(String v) { this.counterpartyId = v; return this; }
        public ListParams q(String v) { this.q = v; return this; }
        public ListParams type(String v) { this.type = v; return this; }
        public ListParams referenceCode(String v) { this.referenceCode = v; return this; }

        Map<String, String> toQuery() {
            return QueryParams.of()
                    .put("start_date", startDate)
                    .put("end_date", endDate)
                    .put("scope", scope)
                    .put("page", page)
                    .put("per_page", perPage)
                    .put("transactions_after_id", transactionsAfterId)
                    .put("transactions_before_id", transactionsBeforeId)
                    .put("transaction_date_after", transactionDateAfter)
                    .put("transaction_date_before", transactionDateBefore)
                    .put("created_after", createdAfter)
                    .put("updated_after", updatedAfter)
                    .put("amount_min", amountMin)
                    .put("amount_max", amountMax)
                    .put("amount", amount)
                    .put("status", status)
                    .put("category_id", categoryId)
                    .put("counterparty_id", counterpartyId)
                    .put("q", q)
                    .put("type", type)
                    .put("reference_code", referenceCode)
                    .build();
        }
    }
}
